src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java

```diff
@@ -672,22 +690,22 @@ public long transferTo(long position, long count,
         long sz = size();
         if (position > sz)
             return 0;
-        int icount = (int)Math.min(count, Integer.MAX_VALUE);
-        if ((sz - position) < icount)
-            icount = (int)(sz - position);
-
-        long n;
-
-        // Attempt a direct transfer, if the kernel supports it
-        if ((n = transferToDirectly(position, icount, target)) >= 0)
-            return n;
 
-        // Attempt a mapped transfer, but only to trusted channel types
-        if ((n = transferToTrustedChannel(position, icount, target)) >= 0)
-            return n;
+        long bytesTransferred = 0L;
+        final long maxTransferSize = maxTransferSize0();
+        while (bytesTransferred < count) {
+            int icount =
+                (int)Math.min(count - bytesTransferred, maxTransferSize);
+            if ((sz - position) < icount)
+                icount = (int)(sz - position);
+            long n = transferTo(position, icount, target);
+            if (n <= 0)
+                break;
+            position += n;
+            bytesTransferred += n;
```

Comment:
I think this patch is problematic for the cases where the second (or subsequent) call to transferTo fails transferToArbitraryChannel has to deal with this by returning the number of bytes transferred before the error.

I think it might be better to just continue with having transferTo map to one syscall where possible. In this case I'd prefer not introduce a loop but instead just use the OS "max transfer size" to cap the number of bytes transferred in one call to this method.

Reply:
In the updated code, the direct transfer case no longer loops and just maps to one syscall. The previously introduced private `transferTo()` method is removed. `transferToArbitraryChannel()` is modified to have a `long`-valued count. `transferToTrustedChannel()` works out of the box as-is and is not changed.

Reply:
The update mostly looks okay, just wondering if could case maxDirectTransferSize to avoid the JNI call every time.

Reply:
I think it could be a constant equal to either the Linux or Windows value, not `Integer.MAX_VALUE`, as that would force `TransmiteFile()` to fail for the `MAX_VALUE` case.